Requirements.md

```diff
@@ -0,0 +1,101 @@
+   
+1. Security is more important than architectural integrity.
+
+1. Architectural integrity is more important than performance.
+   'Architectural Integrity' is intended to denote simplicity/coherence of design. Generally speaking, architectural integrity can be measured in terms of the number of concepts and the complexity of their relationships.
+   
+   The more complex an architecture is, the harder it will be to extend or to modify; this, in turn, inhibits our ability to extend WASM-GC with new features.
+   
+   We prioriterize simple, composeable architectures over solution-oriented architectures. This will help to ensure that WASM-GC can evolove and be part of future WASM evolution.
+   
+   Our reasoning here is that performance is often the result of progressive tuning of an implementation. It can typically be improved over time. On the other hand, architectural deficiencies are very difficult to correct.
+      
+1. Performance is more important than convenience.
+
+   Since WASM-GC is not intended as a user-facing technology, convenience here refers to the ease with which toolchain authors can target WASM-GC. 
+   
+   Convenience for toolchain authors is still important however, as that affects adoption of WASM-GC. 
+
+   
+## Critical Success Factors
+
+A critical success factor is an aspect or property of possible solutions that may or may not be directly focused on the primary objective, but none-the-less is estimated to be crucially important to the success of the effort.
+
+1. Permit ‘cycle detection’ between host structures and language structures; so that such cycles can be collected if there are no other references to them.
+
+   When a WASM-GC module accesses host capabilities, or when a host application access structures from a WASM-GC library, cross references between them are likely to be established. For example, a DOM node may have an event listener that is actually implemented in WASM. The event listener, in turn, may have a reference to the DOM node it is listening to. This represents a cycle between the host and WASM-GC. If there are no other references to this cycles (if, for example, the DOM node is removed) then the entire cycle must be able to be collected.
+
+1. Performance implications of WASM-GC code should be straightforward.
+
+   A general principle behind WASM itself is that performance should not require special implementation techniques on the part of the engine. I.e., it should not be required that the engine discover particular opportunities for improving performance. In particular, it is the responsibility of languages implementers to generate WASM-GC code that embodies performant strategies.
+   
+   For example, if a type cast is potentially eliminatable, then the engine should not be required to discover that fact; rather the toolchain should be able to eliminate the type cast.
+   
+1. Support code migration.
```

Comment:
I agree that it would be useful to have a more detailed picture of what a code migration story would look like.

Reply:
@tlively 

> I agree that it would be useful to have a more detailed picture of what a code migration story would look like.

There is going to have to be a balance, obviously. At one extreme, a single JS expression could be 'replaced' by a call to a WASM function; at the other extreme, the minimum unit of migration is a module.

More likely, the issue will revolve around the kinds of access and the kinds of structure that must be supported.

@brion
>this requires either that Wasm GC objects be implemented as JavaScript objects,

IMPO, if we end up at this place, the entire enterprise risks losing out to another CSF: that WASM be demonstrably better as a target than JS.

Without prejudice, one possible scenario would have JS being able to access WASM objects that are wrapped and that the language running on WASM would be able to access JS objects through a similar wrapping layer that is specfic to the language. This, in turn, likely rules out some of the wilder forms of modification that JS programs are capable of (like messing directly with prototype chains).

Reply:
I think that one way of squaring the interoperability circle is to make explicit (a) that objects may be shared -- in the sense that a change of state by one party is directly observable by other parties and (b) not all the state associated with an entity needs to be observable by all parties.